[ao][sparsity] Implementation of register_layer() and get_mask()

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81814
* #80890
* #80889
* #80888
* __->__ #80887
* #80886

The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

Commit:
[ao][sparsity] Implementation of register_layer() and get_mask()

The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

z-a-f commented on test/ao/sparsity/test_activation_sparsifier.py:
```diff
@@ -57,6 +60,48 @@ def reduce_fn(x):
+
+            return mask
+
+        register_layer1_args = {
+            'layer': model.conv1,
+            'mask_fn': _vanilla_norm_sparsifier
+        }
+        sparse_config_layer1 = {'sparsity_level': 0.3}
+
+        register_layer2_args = {
+            'layer': model.linear1,
+            'features': [0, 10, 234],
+            'feature_dim': 1,
+            'mask_fn': _vanilla_norm_sparsifier
+        }
```
If these are hardcoded, I am not sure we need a dedicated getter method for it? Can't we just use `_vanilla...` without the wrapping method?

macandro96 replied:
done

z-a-f commented on test/ao/sparsity/test_activation_sparsifier.py:
```diff
@@ -78,6 +123,79 @@ def check_constructor(self, activation_sparsifier, model, defaults, sparse_confi
             # all the keys in combined_defaults should be present in sparsifier defaults
             assert config == combined_defaults.get(key, None)
 
+    def check_register_layer(self, activation_sparsifier, defaults, sparse_config, layer_args_list):
```
Again, it looks like the checks are within a generic testcase class, but they are only applicable to your test class. Would be worthhile merging them

macandro96 replied:
done

z-a-f commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -95,24 +99,104 @@ def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=
         self.state: Dict[str, Any] = defaultdict(dict)  # layer name -> mask
 
+    @staticmethod
+    def __safe_rail_checks(args):
```
No "private" classes -- we discourage the use of name mangling. Please, use a single underscore

z-a-f commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -95,24 +99,104 @@ def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=
         Registers a layer for sparsification. The layer should be part of self.model.
         Specifically, registers a pre-forward hook to the layer. The hook will apply the aggregate_fn
         and store the aggregated activations that is input over each step.
 
-        Optionally, accepts custom aggregate, reduce, create_mask functions and features, dimension.
-
         Note::
-            There is no need to pass in the name of the layer as it is automatically computed as per
-            the fqn convention.
+            - There is no need to pass in the name of the layer as it is automatically computed as per
+              the fqn convention.
 
-            All the functions (fn) passed as argument will be called at a dim, feature level.
+            - All the functions (fn) passed as argument will be called at a dim, feature level.
         """
```
Can we move these changes into the previous PR?

z-a-f commented on torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:
```diff
@@ -95,24 +99,104 @@ def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=
+        if features is not None:
+            assert feature_dim is not None, "need feature dim to select features"
+
+        # all the *_fns should be a function
```
nit: "...should be callable"

jerryzh168 commented on test/ao/sparsity/test_activation_sparsifier.py:
```diff
@@ -94,3 +212,25 @@ def test_activation_sparsifier(self):
+            activation_sparsifier.register_layer(**layer_arg, **sparse_config_layer)
+
+        # check if things are registered correctly
+        self.check_register_layer(activation_sparsifier, defaults, sparse_config, layer_args_list)
```
also I'm wondering if we can separate these steps into multiple tests to keep the test itself shorter

macandro96 replied:
Since we're testing the complete workflow here, the state of sparsifier will change at each step. So, we have one single test-case here.

jerryzh168 commented on test/ao/sparsity/test_activation_sparsifier.py:
```diff
@@ -78,6 +123,79 @@ def check_constructor(self, activation_sparsifier, model, defaults, sparse_confi
+            sparse_config_actual.update(sparse_config_layer)
+
+            name = module_to_fqn(activation_sparsifier.model, layer_arg['layer'])
+            assert data_groups[name]['layer'] == layer_arg['layer']
```
also I think ideally we should focus on testing the functionality instead of implementation details, it looks like we are testing implementation details here

macandro96 replied:
removed such tests

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Implementation of register_layer() and get_mask()"


The register_layer() attaches a pre-forward hook to the layer to aggregate
activations over time. The mask shape is also inferred here.

The get_mask() returns the computed mask associated to the attached layer.
The mask is
    - a torch tensor is features for that layer is None.
    - a list of torch tensors for each feature, otherwise

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g
